### PR TITLE
[Merged by Bors] - feat: Valuative criterion for properness.

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/ValuativeCriterion.lean [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+/-!
+# Valuative criterion
+
+## Main results
+
+- `AlgebraicGeometry.UniversallyClosed.eq_valuativeCriterion`:
+  A morphism is universally closed if and only if
+  it is quasi-compact and satisfies the existence part of the valuative criterion.
+- `AlgebraicGeometry.UniversallyClosed.eq_valuativeCriterion`:
```
Comment (jcommelin) [inline Mathlib/AlgebraicGeometry/ValuativeCriterion.lean:19]: I guess this should be
```suggestion
- `AlgebraicGeometry.IsSeparated.eq_valuativeCriterion`:
```

File: Mathlib/AlgebraicGeometry/ValuativeCriterion.lean [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+  A morphism is separated if and only if
+  it is quasi-separated and satisfies the uniqueness part of the valuative criterion.
+- `AlgebraicGeometry.UniversallyClosed.eq_valuativeCriterion`:
```
Comment (jcommelin) [inline Mathlib/AlgebraicGeometry/ValuativeCriterion.lean:22]: And this
```suggestion
- `AlgebraicGeometry.IsProper.eq_valuativeCriterion`:
```

File: Mathlib/AlgebraicGeometry/ValuativeCriterion.lean [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+
+/-- A morphism `f : X ⟶ Y` satisfies the uniqueness part of the valuative criterion if
+every valuative commutative square over `f` has at most one lift. -/
+def ValuativeCriterion.Uniqueness : MorphismProperty Scheme :=
+  fun _ _ f ↦ ∀ S : ValuativeCommSq f, Subsingleton S.commSq.LiftStruct
+
+/-- A morphism `f : X ⟶ Y` satisfies the valuative criterion if
+every valuative commutative square over `f` has a unique lift. -/
+def ValuativeCriterion : MorphismProperty Scheme :=
+  fun _ _ f ↦ ∀ S : ValuativeCommSq f, Nonempty (Unique (S.commSq.LiftStruct))
```
Comment (jcommelin) [inline Mathlib/AlgebraicGeometry/ValuativeCriterion.lean:85]: Please add a little bit of API:
- projections `.existence` and `.uniqueness`,
- and `iff` that shows this is equivalent to `Existence + Uniqueness`.

**Review by jcommelin — COMMENTED** [review]
Very nice!

File: Mathlib/AlgebraicGeometry/ValuativeCriterion.lean (hunk 3 above)
Comment (erdOne) [reply Mathlib/AlgebraicGeometry/ValuativeCriterion.lean:85]: Done.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**